src/btree/bt_sync.c

```diff
@@ -115,129 +115,22 @@ __sync_dup_walk(WT_SESSION_IMPL *session, WT_REF *walk, uint32_t flags, WT_REF *
-
-    /* Ignore root pages as they can never be deleted. */
-    if (__wt_ref_is_root(ref)) {
-        __wt_verbose_debug2(
-          session, WT_VERB_CHECKPOINT_CLEANUP, "%p: skipping root page", (void *)ref);
-        return (0);
-    }
-
-    /* Ignore internal pages, these are taken care of during reconciliation. */
-    if (F_ISSET(ref, WT_REF_FLAG_INTERNAL)) {
-        __wt_verbose_debug2(session, WT_VERB_CHECKPOINT_CLEANUP,
-          "%p: skipping internal page with parent: %p", (void *)ref, (void *)ref->home);
-        return (0);
-    }
-
-    /* Lock the WT_REF. */
-    WT_REF_LOCK(session, ref, &previous_state);
-
-    /* Skip non obsolete deleted pages or remove obsolete deleted pages. */
-    if (previous_state == WT_REF_DELETED) {
-        page_del = ref->page_del;
-        if (page_del == NULL ||
-          __wt_txn_visible_all(session, page_del->txnid, page_del->durable_timestamp)) {
-            ret = __wt_page_parent_modify_set(session, ref, true);
-            WT_REF_UNLOCK(ref, previous_state);
-            WT_RET(ret);
-            __wt_verbose(session, WT_VERB_CHECKPOINT_CLEANUP,
-              "%p: marking obsolete deleted page parent dirty", (void *)ref);
-        } else {
-            __wt_verbose(
-              session, WT_VERB_CHECKPOINT_CLEANUP, "%p: skipping deleted page", (void *)ref);
-            WT_REF_UNLOCK(ref, previous_state);
-        }
-        return (0);
-    }
-
-    /*
-     * If the page is on-disk and obsolete, mark the page as deleted and also set the parent page as
-     * dirty. This is to ensure the parent is written during the checkpoint and the child page
-     * discarded.
-     */
-    newest_stop_ts = WT_TS_NONE;
-    newest_stop_durable_ts = WT_TS_NONE;
-    newest_stop_txn = WT_TXN_NONE;
-    obsolete = false;
-    ovfl_items = false;
-    if (previous_state == WT_REF_DISK) {
-        /*
-         * There should be an address, but simply skip any page where we don't find one. Also skip
-         * the pages that have overflow keys as part of fast delete flow. These overflow keys pages
-         * are handled as an in-memory obsolete page flow.
-         */
-        if (__wt_ref_addr_copy(session, ref, &addr) && addr.type == WT_ADDR_LEAF_NO) {
-            /*
-             * Max stop timestamp is possible only when the prepared update is written to the data
-             * store.
-             */
-            newest_stop_ts = addr.ta.newest_stop_ts;
-            newest_stop_durable_ts =
-              addr.ta.newest_stop_ts == WT_TS_MAX ? WT_TS_MAX : addr.ta.newest_stop_durable_ts;
-            newest_stop_txn = addr.ta.newest_stop_txn;
-            obsolete = __wt_txn_visible_all(session, newest_stop_txn, newest_stop_durable_ts);
-        }
-
-        if (obsolete) {
-            if ((ret = __wt_page_parent_modify_set(session, ref, true)) != 0)
-                WT_REF_UNLOCK(ref, previous_state);
-            else
-                WT_REF_UNLOCK(ref, WT_REF_DELETED);
```

Comment:
I feel like this transition is gone now. I don't see any path in the new code where a WT_REF_DISK becomes a WT_REF_DELETED. Is that deliberate?

Reply:
Yes, I missed it in one of the commits, but I fixed it later to keep the transition. Without this transition, the checkpoint cleanup doesn't free any obsolete ref.